CMIP7/esm1p6/atmosphere/ghg/cmip7_HI_ghg_generate.py

```diff
@@ -58,9 +58,50 @@ def load_cmip7_hi_ghg_mmr(args, ghg):
     return ghg_mmr_list
 
 
-def cmip7_hi_ghg_patch(ghg_mmr_dict):
+def read_namelists_lines_up_to(namelists_filepath, exclude_str):
     """
-    Patch the greenhouse gas variables in the clmchfcg namelist
+    Read lines from namelists_filepath up to but not including
+    the line containing exclude_str.
+    """
+    if not namelists_filepath.exists():
+        raise FileNotFoundError(
+            f"Namelist file {namelists_filepath} does not exist"
+        )
+    # Read the atmosphere/namelists file up to but not including
+    # the clmchfcg namelist.
+    namelists_str = ""
+    with open(namelists_filepath) as namelists_file:
+        namelists_line = namelists_file.readline()
+        while namelists_line and (
+            "&" + exclude_str not in namelists_line.lower()
+        ):
+            namelists_str += namelists_line
+            namelists_line = namelists_file.readline()
+    return namelists_str
```

Comment:
The opened file handle `namelists_file` is an iterable that already contains each line of the file. 
So this could be simplified by iterating over `namelists_file` and stopping when the exclusion string is found, appending each line to the overall result.
Also, it is advised to set the whole "exclude_str" into a variable first, and then use it directly, without hardcoding the `&` within the call: 
```suggestion
    exclude_str = "&" + exclude_str
    namelist=[]
    with open(namelists_filepath) as namelists_file:
        for line in namelists_file:
            if exclude_str in line:
                break
            namelist.append(line)
    return "".join(namelist)
```